iocell/src/main/scala/chisel/IOCell.scala

```diff
@@ -0,0 +1,188 @@
+// See LICENSE for license details
+
+package barstools.iocell.chisel
+
+import chisel3._
+import chisel3.util.{Cat, HasBlackBoxResource}
+import chisel3.experimental.{Analog, DataMirror, IO}
+
+class AnalogIOCellBundle extends Bundle {
+  val pad = Analog(1.W)
+  val core = Analog(1.W)
+}
+
+class DigitalGPIOCellBundle extends Bundle {
+  val pad = Analog(1.W)
+  val i = Output(Bool())
+  val ie = Input(Bool())
+  val o = Input(Bool())
+  val oe = Input(Bool())
+}
+
+class DigitalOutIOCellBundle extends Bundle {
+  val pad = Output(Bool())
+  val o = Input(Bool())
+  val oe = Input(Bool())
+}
+
+class DigitalInIOCellBundle extends Bundle {
+  val pad = Input(Bool())
+  val i = Output(Bool())
+  val ie = Input(Bool())
+}
+
+abstract class IOCell extends BlackBox with HasBlackBoxResource
+
+abstract class AnalogIOCell extends IOCell {
+  val io: AnalogIOCellBundle
+}
+
+abstract class DigitalGPIOCell extends IOCell {
+  val io: DigitalGPIOCellBundle
+}
+
+abstract class DigitalInIOCell extends IOCell {
+  val io: DigitalInIOCellBundle
+}
+
+abstract class DigitalOutIOCell extends IOCell {
+  val io: DigitalOutIOCellBundle
+}
+
+class ExampleAnalogIOCell extends AnalogIOCell {
+  val io = IO(new AnalogIOCellBundle)
+  addResource("/barstools/iocell/vsrc/IOCell.v")
+}
+
+class ExampleDigitalGPIOCell extends DigitalGPIOCell {
+  val io = IO(new DigitalGPIOCellBundle)
+  addResource("/barstools/iocell/vsrc/IOCell.v")
+}
+
+class ExampleDigitalInIOCell extends DigitalInIOCell {
+  val io = IO(new DigitalInIOCellBundle)
+  addResource("/barstools/iocell/vsrc/IOCell.v")
+}
+
+class ExampleDigitalOutIOCell extends DigitalOutIOCell {
+  val io = IO(new DigitalOutIOCellBundle)
+  addResource("/barstools/iocell/vsrc/IOCell.v")
+}
+
+object IOCell {
+
+  def exampleAnalog() = Module(new ExampleAnalogIOCell)
+  def exampleGPIO() = Module(new ExampleDigitalGPIOCell)
+  def exampleInput() = Module(new ExampleDigitalInIOCell)
+  def exampleOutput() = Module(new ExampleDigitalOutIOCell)
+
+  def generateIOFromSignal[T <: Data](coreSignal: T, name: Option[String] = None,
+    inFn: () => DigitalInIOCell = IOCell.exampleInput,
+    outFn: () => DigitalOutIOCell = IOCell.exampleOutput,
+    anaFn: () => AnalogIOCell = IOCell.exampleAnalog): (T, Seq[IOCell]) =
+  {
+    val padSignal = IO(DataMirror.internal.chiselTypeClone[T](coreSignal))
+    val iocells = IOCell.generateFromSignal(coreSignal, padSignal, name, inFn, outFn, anaFn)
+    (padSignal, iocells)
+  }
+
+  def generateFromSignal[T <: Data](coreSignal: T, padSignal: T, name: Option[String] = None,
+    inFn: () => DigitalInIOCell = IOCell.exampleInput,
+    outFn: () => DigitalOutIOCell = IOCell.exampleOutput,
+    anaFn: () => AnalogIOCell = IOCell.exampleAnalog): Seq[IOCell] =
+  {
+    coreSignal match {
+      case coreSignal: Analog => {
+        if (coreSignal.getWidth == 0) {
```

Comment:
Is it possible to hoist this compare out of all the individual cases?

Reply:
No- the comparison is handled differently depending on the underlying type. For `Analog`, specifically, widths of 0 or > 1 are impossible to support, so we have to assert. For most other things, any widths are fine.

Reply:
But is returning an empty Seq acceptable for all zero width signals?

Reply:
Unfortunately no, you have to handle the `Bits` case specially (scroll down to bits to see the `0.U` dummy assignment)